Sources/DynamicColor+RGBA.swift

```diff
@@ -62,14 +62,11 @@ public extension DynamicColor {
 
       return (r, g, b, a)
     #elseif os(OSX)
-      if isEqual(DynamicColor.black) {
-        return (0, 0, 0, 0)
-      }
-      else if isEqual(DynamicColor.white) {
-        return (1, 1, 1, 1)
+      guard let rgbaColor = self.usingColorSpace(.deviceRGB) else {
+        fatalError("Could not convert color to RGBA.")
```

Comment:
What do you think about using an `assert` or `assertFailure` instead of the `fatalError`? In some use case we may not know which kind of colour we manipulate and it can be harmful to crash the app.

Reply:
The current (i.e. `master`) implementation effectively looks like this, when on macOS:

```swift
final func toRGBAComponents() -> (r: CGFloat, g: CGFloat, b: CGFloat, a: CGFloat) {
    var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
    self.getRed(&r, green: &g, blue: &b, alpha: &a)
    return (r, g, b, a)
}
```

Calling this function on anything, that's not an RGBA color, while on macOS, crashes:

<details>
<summary>NSInvalidArgumentException</summary>

```
caught "NSInvalidArgumentException", "*** -getRed:green:blue:alpha: not valid for the NSColor Generic Gray Gamma 2.2 Profile colorspace 0.42 1; need to first convert colorspace."
(
	0   CoreFoundation                      0x00007fff3b1518ab __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff71887805 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff3b151701 +[NSException raise:format:] + 189
	3   DynamicColorTests                   0x0000000102632373 $sSo7NSColorC12DynamicColorE16toRGBAComponents12CoreGraphics7CGFloatV1r_AG1gAG1bAG1atyF + 115
	4   DynamicColorTests                   0x0000000102607dcd $s17DynamicColorTests0aB9RGBATestsC20testToRGBAComponentsyyF + 1949
	5   DynamicColorTests                   0x0000000102608c3b $s17DynamicColorTests0aB9RGBATestsC20testToRGBAComponentsyyFTo + 43
	6   CoreFoundation                      0x00007fff3b0b6fec __invoking___ + 140
	7   CoreFoundation                      0x00007fff3b0b6e83 -[NSInvocation invoke] + 305
	8   XCTest                              0x00000001003d1cc6 __24-[XCTestCase invokeTest]_block_invoke_2.214 + 65
	9   XCTest                              0x000000010045fcb7 -[XCTMemoryChecker _assertInvalidObjectsDeallocatedAfterScope:] + 51
	10  XCTest                              0x00000001003dcbae -[XCTestCase assertInvalidObjectsDeallocatedAfterScope:] + 116
	11  XCTest                              0x00000001003d1c56 __24-[XCTestCase invokeTest]_block_invoke.208 + 204
	12  XCTest                              0x0000000100445da0 -[XCTestCase(Failures) performFailableBlock:testCaseRun:shouldInterruptTest:] + 57
	13  XCTest                              0x0000000100445cbd -[XCTestCase(Failures) _performTurningExceptionsIntoFailuresInterruptAfterHandling:block:] + 96
	14  XCTest                              0x00000001003d1707 __24-[XCTestCase invokeTest]_block_invoke + 1120
	15  XCTest                              0x00000001003d1200 -[XCTestCase testContextPerformInScope:] + 184
	16  XCTest                              0x00000001003d129a -[XCTestCase invokeTest] + 137
	17  XCTest                              0x00000001003d3079 __26-[XCTestCase performTest:]_block_invoke_2 + 43
	18  XCTest                              0x0000000100445da0 -[XCTestCase(Failures) performFailableBlock:testCaseRun:shouldInterruptTest:] + 57
	19  XCTest                              0x0000000100445cbd -[XCTestCase(Failures) _performTurningExceptionsIntoFailuresInterruptAfterHandling:block:] + 96
	20  XCTest                              0x00000001003d2f90 __26-[XCTestCase performTest:]_block_invoke.342 + 88
	21  XCTest                              0x0000000100457380 +[XCTContext runInContextForTestCase:block:] + 219
	22  XCTest                              0x00000001003d2703 -[XCTestCase performTest:] + 668
	23  XCTest                              0x00000001004248e3 -[XCTest runTest] + 57
	24  XCTest                              0x00000001003ccda3 __27-[XCTestSuite performTest:]_block_invoke + 365
	25  XCTest                              0x00000001003cc4cd -[XCTestSuite _performProtectedSectionForTest:testSection:] + 54
	26  XCTest                              0x00000001003cc7ca -[XCTestSuite performTest:] + 355
	27  XCTest                              0x00000001004248e3 -[XCTest runTest] + 57
	28  XCTest                              0x00000001003ccda3 __27-[XCTestSuite performTest:]_block_invoke + 365
	29  XCTest                              0x00000001003cc4cd -[XCTestSuite _performProtectedSectionForTest:testSection:] + 54
	30  XCTest                              0x00000001003cc7ca -[XCTestSuite performTest:] + 355
	31  XCTest                              0x00000001004248e3 -[XCTest runTest] + 57
	32  XCTest                              0x00000001003ccda3 __27-[XCTestSuite performTest:]_block_invoke + 365
	33  XCTest                              0x00000001003cc4cd -[XCTestSuite _performProtectedSectionForTest:testSection:] + 54
	34  XCTest                              0x00000001003cc7ca -[XCTestSuite performTest:] + 355
	35  XCTest                              0x00000001004248e3 -[XCTest runTest] + 57
	36  XCTest                              0x0000000100471886 __44-[XCTTestRunSession runTestsAndReturnError:]_block_invoke + 171
	37  XCTest                              0x0000000100471982 __44-[XCTTestRunSession runTestsAndReturnError:]_block_invoke.84 + 118
	38  XCTest                              0x00000001003f36a7 -[XCTestObservationCenter _observeTestExecutionForBlock:] + 594
	39  XCTest                              0x0000000100471644 -[XCTTestRunSession runTestsAndReturnError:] + 623
	40  XCTest                              0x00000001003afa40 -[XCTestDriver runTestsAndReturnError:] + 458
	41  XCTest                              0x00000001004536c8 _XCTestMain + 2473
	42  xctest                              0x00000001000020ca main + 245
	43  libdyld.dylib                       0x00007fff72bf57fd start + 1
	44  ???                                 0x0000000000000005 0x0 + 5
)
```
</details>

On iOS it doesn't crash, as UIColor performs a color space conversion, if necessary. NSColor  however doesn't. It just crashes.

The only reason you didn't notice it yet, was that you weren't using  any non-RGBA colors with that method. As such you were creating black and white colors via `DynamicColor(red: 0, green: 0, blue: 0, alpha: 0)` (which creates a RGBA color), instead of `DynamicColor(white, alpha: 0)` (which creates a grayscale color), e.g. My tests for https://github.com/yannickl/DynamicColor/pull/55 however do use a true grayscale color and that's how I found the bug.

The only non-crashing alternative I can think of would be to bail out and return incorrect RGBA components, instead, by changing …

```swift
guard let rgbaColor = self.usingColorSpace(.deviceRGB) else {
    fatalError("Could not convert color to RGBA.")
}
```

… to …

```swift
guard let rgbaColor = self.usingColorSpace(.deviceRGB) else {
    assertionFailure("Could not convert color to RGBA.")
    return (r, g, b, a)
}
```

… which would have the downside of silently failing. Good luck debugging, trying to  figure out why all kinds of colors return black on a Mac, while working just fine on iOS.

Reply:
Anyway, I'm happy to change the PR to whatever choice you make for this. :)

Reply:
Thanks for these great explanations! If it was failing before I will let your this as is. It seems fair enough.